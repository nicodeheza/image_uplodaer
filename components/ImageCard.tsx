export default function ImageCard() {
	return (
		<div>
			<div></div>
		</div>
	);
}
