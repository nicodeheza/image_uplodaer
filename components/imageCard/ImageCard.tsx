import Image from "next/image";
import {Dispatch, useLayoutEffect, useState, SetStateAction} from "react";
import CheckIcon from "../icons/CheckIcon";
import styles from "./imageCard.module.css";

export interface ImageCardProps {
	imageName: string;
	setShowCopy: Dispatch<SetStateAction<boolean>>;
}

export default function ImageCard({imageName, setShowCopy}: ImageCardProps) {
	const [imageUrl, setImageUrl] = useState("");

	useLayoutEffect(() => {
		setImageUrl(
			`${window.location.protocol}//${window.location.host}/uploads/${imageName}`
		);
	}, [imageName]);

	async function copyLink(link: string) {
		try {
			await navigator.clipboard.writeText(link);
			setShowCopy(true);
		} catch (error) {
			alert("Error copying image url");
			console.log(error);
		}
	}
	return (
		<div className={styles.imageCard}>
			<div>
				<CheckIcon />
			</div>
			<h1>Uploaded Successfully!</h1>
			<Image
				width={338}
				height={224.4}
				src={`/uploads/${imageName}`}
				alt="loaded Image"
				objectFit="contain"
				priority
			/>
			<div className={styles.linkContainer}>
				<div className={styles.urlContainer}>
					<p>{imageUrl}</p>
				</div>
				<button onClick={() => copyLink(imageUrl)}>Copy Link</button>
			</div>
		</div>
	);
}
