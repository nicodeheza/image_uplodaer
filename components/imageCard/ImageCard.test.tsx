import ImageCard from "./ImageCard";
import {fireEvent, render} from "@testing-library/react";
import renderer from "react-test-renderer";

Object.assign(navigator, {
	clipboard: {
		writeText: () => {}
	}
});

const imageName = "Chuck-Norris.jpg";
const imageUrlRegExp = /^(http|https):\/\/[\w\W]+\/uploads\/Chuck-Norris.jpg$/;

describe("ImageCard component", () => {
	let imageUrl: string;
	it("image must be rendered successfully", () => {
		const {getByAltText} = render(<ImageCard imageName={imageName} />);
		const image = getByAltText("loaded Image");
		const src = image.getAttribute("src");

		const isImage = /(?=.*Chuck-Norris\.jpg)(?=.*uploads).*/.test(src!);
		expect(isImage).toBeTruthy();
	});
	it("image url must be displayed", () => {
		const {getByText} = render(<ImageCard imageName={imageName} />);
		const url = getByText(imageUrlRegExp);
		imageUrl = url.textContent!;

		expect(url).toBeTruthy();
	});
	it("when copy button is clicked the url must be copied to the clipboard", () => {
		const {getByText} = render(<ImageCard imageName={imageName} />);
		const button = getByText("Copy Link");
		const spy = jest.spyOn(navigator.clipboard, "writeText");

		expect(button).toBeTruthy();
		fireEvent.click(button);

		expect(spy).toHaveBeenCalledWith(imageUrl);
	});
	it("snapshot", () => {
		const tree = renderer.create(<ImageCard imageName={imageName} />).toJSON();
		expect(tree).toMatchSnapshot();
	});
});
