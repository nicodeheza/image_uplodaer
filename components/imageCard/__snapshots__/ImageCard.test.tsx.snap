// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageCard component snapshot 1`] = `
<div
  className="imageCard"
>
  <div>
    <svg
      fill="#219653"
      height="48"
      width="48"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M21.1 32.95 35.05 19l-2.2-2.2L21.1 28.55l-5.95-5.95-2.2 2.2Zm2.9 10.8q-4.05 0-7.65-1.575-3.6-1.575-6.275-4.25Q7.4 35.25 5.825 31.65 4.25 28.05 4.25 24q0-4.1 1.575-7.7 1.575-3.6 4.25-6.275Q12.75 7.35 16.35 5.8q3.6-1.55 7.65-1.55 4.1 0 7.7 1.55 3.6 1.55 6.275 4.225Q40.65 12.7 42.2 16.3q1.55 3.6 1.55 7.7 0 4.05-1.55 7.65-1.55 3.6-4.225 6.275Q35.3 40.6 31.7 42.175 28.1 43.75 24 43.75Z"
      />
    </svg>
  </div>
  <h1>
    Uploaded Successfully!
  </h1>
  <span
    style={
      Object {
        "background": "none",
        "border": 0,
        "boxSizing": "border-box",
        "display": "inline-block",
        "height": "initial",
        "margin": 0,
        "maxWidth": "100%",
        "opacity": 1,
        "overflow": "hidden",
        "padding": 0,
        "position": "relative",
        "width": "initial",
      }
    }
  >
    <span
      style={
        Object {
          "background": "none",
          "border": 0,
          "boxSizing": "border-box",
          "display": "block",
          "height": "initial",
          "margin": 0,
          "maxWidth": "100%",
          "opacity": 1,
          "padding": 0,
          "width": "initial",
        }
      }
    >
      <img
        alt=""
        aria-hidden={true}
        src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%27338%27%20height=%27224.4%27/%3e"
        style={
          Object {
            "background": "none",
            "border": 0,
            "display": "block",
            "height": "initial",
            "margin": 0,
            "maxWidth": "100%",
            "opacity": 1,
            "padding": 0,
            "width": "initial",
          }
        }
      />
    </span>
    <img
      alt="loaded Image"
      data-nimg="intrinsic"
      decoding="async"
      onError={[Function]}
      onLoad={[Function]}
      src="/_next/image?url=%2Fuploads%2FChuck-Norris.jpg&w=750&q=75"
      srcSet="/_next/image?url=%2Fuploads%2FChuck-Norris.jpg&w=384&q=75 1x, /_next/image?url=%2Fuploads%2FChuck-Norris.jpg&w=750&q=75 2x"
      style={
        Object {
          "border": "none",
          "bottom": 0,
          "boxSizing": "border-box",
          "display": "block",
          "height": 0,
          "left": 0,
          "margin": "auto",
          "maxHeight": "100%",
          "maxWidth": "100%",
          "minHeight": "100%",
          "minWidth": "100%",
          "objectFit": undefined,
          "objectPosition": undefined,
          "padding": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": 0,
        }
      }
    />
  </span>
  <div
    className="linkContainer"
  >
    <div
      className="urlContainer"
    >
      <p>
        http://localhost/uploads/Chuck-Norris.jpg
      </p>
    </div>
    <button
      onClick={[Function]}
    >
      Copy Link
    </button>
  </div>
</div>
`;
